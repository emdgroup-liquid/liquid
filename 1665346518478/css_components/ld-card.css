.ld-card,:host{--ld-card-bg-col:var(--ld-col-wht);--ld-card-border-radius:var(--ld-br-l);--ld-card-padding:var(--ld-sp-32);--ld-card-padding-sm:var(--ld-sp-16);-webkit-appearance:none;appearance:none;background-color:var(--ld-card-bg-col);border:none;border-radius:var(--ld-card-border-radius);box-sizing:border-box;color:inherit;display:inline-grid;justify-items:flex-start;list-style:none;padding:var(--ld-card-padding);position:relative;text-decoration:none}.ld-card:after,.ld-card:before,:host:after,:host:before{border-radius:inherit;box-shadow:var(--ld-shadow-stacked);content:"";inset:0;pointer-events:none;position:absolute;transition:opacity .2s ease-out}.ld-card:after,:host:after{opacity:0}.ld-card--interactive-sticky:after,.ld-card--sticky:before,:host(.ld-card--interactive-sticky):after,:host(.ld-card--sticky):before{box-shadow:var(--ld-shadow-sticky)}.ld-card--hover:before,.ld-card--interactive-hover:after,:host(.ld-card--hover):before,:host(.ld-card--interactive-hover):after{box-shadow:var(--ld-shadow-hover)}.ld-card--active:before,.ld-card--interactive-active:after,:host(.ld-card--active):before,:host(.ld-card--interactive-active):after{box-shadow:var(--ld-shadow-active)}.ld-card--interactive-active:focus-within:before,.ld-card--interactive-active:hover:before,.ld-card--interactive-hover:focus-within:before,.ld-card--interactive-hover:hover:before,.ld-card--interactive-stacked:focus-within:before,.ld-card--interactive-stacked:hover:before,.ld-card--interactive-sticky:focus-within:before,.ld-card--interactive-sticky:hover:before,:host(.ld-card--interactive-active:focus-within):before,:host(.ld-card--interactive-active:hover):before,:host(.ld-card--interactive-hover:focus-within):before,:host(.ld-card--interactive-hover:hover):before,:host(.ld-card--interactive-stacked:focus-within):before,:host(.ld-card--interactive-stacked:hover):before,:host(.ld-card--interactive-sticky:focus-within):before,:host(.ld-card--interactive-sticky:hover):before{opacity:0}.ld-card--interactive-active:focus-within:after,.ld-card--interactive-active:hover:after,.ld-card--interactive-hover:focus-within:after,.ld-card--interactive-hover:hover:after,.ld-card--interactive-stacked:focus-within:after,.ld-card--interactive-stacked:hover:after,.ld-card--interactive-sticky:focus-within:after,.ld-card--interactive-sticky:hover:after,:host(.ld-card--interactive-active:focus-within):after,:host(.ld-card--interactive-active:hover):after,:host(.ld-card--interactive-hover:focus-within):after,:host(.ld-card--interactive-hover:hover):after,:host(.ld-card--interactive-stacked:focus-within):after,:host(.ld-card--interactive-stacked:hover):after,:host(.ld-card--interactive-sticky:focus-within):after,:host(.ld-card--interactive-sticky:hover):after{opacity:1}.ld-card--sm,:host(.ld-card--sm){padding:var(--ld-card-padding-sm)}