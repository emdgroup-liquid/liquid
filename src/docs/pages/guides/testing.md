---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

