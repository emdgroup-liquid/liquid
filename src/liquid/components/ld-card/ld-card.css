.ld-card {
  --ld-card-bg-col: var(--ld-col-wht);
  --ld-card-border-radius: var(--ld-br-l);
  --ld-card-padding: var(--ld-sp-32);
  --ld-card-padding-sm: var(--ld-sp-16);

  appearance: none;
  border: none;
  box-sizing: border-box;
  background-color: var(--ld-card-bg-col);
  border-radius: var(--ld-card-border-radius);
  color: inherit;
  display: inline-grid;
  justify-items: flex-start;
  list-style: none;
  padding: var(--ld-card-padding);
  position: relative;
  text-decoration: none;

  &::before,
  &::after {
    content: '';
    border-radius: inherit;
    box-shadow: var(--ld-shadow-stacked);
    inset: 0;
    pointer-events: none;
    position: absolute;
    transition: opacity 0.2s ease-out;
  }

  &::after {
    opacity: 0;
  }

  &--sticky::before,
  &--interactive-sticky::after {
    box-shadow: var(--ld-shadow-sticky);
  }
  &--hover::before,
  &--interactive-hover::after {
    box-shadow: var(--ld-shadow-hover);
  }
  &--active::before,
  &--interactive-active::after {
    box-shadow: var(--ld-shadow-active);
  }

  &--interactive-stacked,
  &--interactive-sticky,
  &--interactive-hover,
  &--interactive-active {
    &:hover,
    &:focus-within {
      &::before {
        opacity: 0;
      }
      &::after {
        opacity: 1;
      }
    }
  }

  &--sm {
    padding: var(--ld-card-padding-sm);
  }
}
