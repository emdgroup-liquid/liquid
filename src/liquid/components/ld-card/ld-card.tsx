import '../../components' // type definitions for type checks and intelliSense
import { Component, h, Host, Prop } from '@stencil/core'
import { getClassNames } from '../../utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 */
@Component({
  tag: 'ld-card',
  styleUrl: 'ld-card.css',
  shadow: true,
})
export class LdCard {
  /** Simulates card elevation by setting the size of the card box shadow. */
  @Prop() shadow: 'active' | 'hover' | 'stacked' | 'sticky' = 'stacked'

  /**
   * Adds hover and focus-within states using an elevation transition from
   * `shadow` (see above) to `shadowInteractive`.
   */
  @Prop() shadowInteractive?: 'active' | 'hover' | 'stacked' | 'sticky'

  /** The size prop effects the card padding only. */
  @Prop() size?: 'sm'

  render() {
    const cl = getClassNames([
      'ld-card',
      this.size && `ld-card--${this.size}`,
      this.shadow && `ld-card--${this.shadow}`,
      this.shadowInteractive &&
        `ld-card--interactive-${this.shadowInteractive}`,
    ])

    return (
      <Host class={cl}>
        <slot></slot>
      </Host>
    )
  }
}
