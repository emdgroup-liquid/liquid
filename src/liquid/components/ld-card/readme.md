---
eleventyNavigation:
  key: Card
  parent: Components
layout: layout.njk
title: Card
permalink: components/ld-card/
---

<link rel="stylesheet" href="{{ env.base }}/{{ buildstamp }}css_components/ld-card.css">
<link rel="stylesheet" href="{{ env.base }}/{{ buildstamp }}css_components/ld-typo.css">

# ld-card

The `ld-card` component can be used to list content in visually clearly separated and highlighted areas.

Additionally, there is the [`ld-card-stack`](components/ld-card/ld-card-stack/) component, which allows you to display multiple `ld-card`s in a stack.

---

## Default

{% example %}
<ld-card>
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<!-- React component -->

<LdCard>
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<!-- CSS component -->

<div class="ld-card">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>
{% endexample %}

## Size

The `size` prop effects the card padding only.

{% example %}
<ld-card>
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<ld-card size="sm">
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<!-- React component -->

<LdCard>
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<LdCard size="sm">
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<!-- CSS component -->

<div class="ld-card">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>

<div class="ld-card ld-card--sm">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>
{% endexample %}

## Shadow

{% example %}
<ld-card>
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<ld-card shadow="active">
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<ld-card shadow="hover">
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<ld-card shadow="sticky">
  <ld-typo>Lorem ipsum dolor sit amet.</ld-typo>
</ld-card>

<!-- React component -->

<LdCard>
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<LdCard shadow="active">
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<LdCard shadow="hover">
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<LdCard shadow="sticky">
  <LdTypo>Lorem ipsum dolor sit amet.</LdTypo>
</LdCard>

<!-- CSS component -->

<div class="ld-card">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>

<div class="ld-card ld-card--active">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>

<div class="ld-card ld-card--hover">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>

<div class="ld-card ld-card--sticky">
  <p class="ld-typo ld-typo--body-m">Lorem ipsum dolor sit amet.</p>
</div>
{% endexample %}

## Shadow interactive

Use the `shadow-interactive` prop for a transition to a different shadow on hover and focus-within.

{% example %}
<ld-card shadow-interactive="sticky">
  <ld-typo style="margin-bottom: 1rem">
    Lorem ipsum dolor sit amet.
  </ld-typo>
  <ld-button mode="highlight">Click me</ld-button>
</ld-card>

<ld-card shadow="sticky" shadow-interactive="stacked">
  <ld-typo style="margin-bottom: 1rem">
    Lorem ipsum dolor sit amet.
  </ld-typo>
  <ld-button mode="highlight">Click me</ld-button>
</ld-card>

<!-- React component -->

<LdCard shadowInteractive="sticky">
  <LdTypo style={ { marginBottom: '1rem' } }>
    Lorem ipsum dolor sit amet.
  </LdTypo>
  <LdButton mode="highlight">Click me</LdButton>
</LdCard>

<LdCard shadow="sticky" shadowInteractive="stacked">
  <LdTypo style={ { marginBottom: '1rem' } }>
    Lorem ipsum dolor sit amet.
  </LdTypo>
  <LdButton mode="highlight">Click me</LdButton>
</LdCard>

<!-- CSS component -->

<div class="ld-card ld-card--interactive-sticky">
  <p class="ld-typo ld-typo--body-m" style="margin-bottom: 1rem">
    Lorem ipsum dolor sit amet.
  </p>
  <button class="ld-button ld-button--highlight">
    Click me
  </button>
</div>

<div class="ld-card ld-card--sticky ld-card--interactive-stacked">
  <p class="ld-typo ld-typo--body-m" style="margin-bottom: 1rem">
    Lorem ipsum dolor sit amet.
  </p>
  <button class="ld-button ld-button--highlight">
    Click me
  </button>
</div>
{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property            | Attribute            | Description                                                                                                        | Type                                           | Default     |
| ------------------- | -------------------- | ------------------------------------------------------------------------------------------------------------------ | ---------------------------------------------- | ----------- |
| `key`               | `key`                | for tracking the node's identity when working with lists                                                           | `string \| number`                             | `undefined` |
| `ref`               | `ref`                | reference to component                                                                                             | `any`                                          | `undefined` |
| `shadow`            | `shadow`             | Simulates card elevation by setting the size of the card box shadow.                                               | `"active" \| "hover" \| "stacked" \| "sticky"` | `'stacked'` |
| `shadowInteractive` | `shadow-interactive` | Adds hover and focus-within states using an elevation transition from `shadow` (see above) to `shadowInteractive`. | `"active" \| "hover" \| "stacked" \| "sticky"` | `undefined` |
| `size`              | `size`               | The size prop effects the card padding only.                                                                       | `"sm"`                                         | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
