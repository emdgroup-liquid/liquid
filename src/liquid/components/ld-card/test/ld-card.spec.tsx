import { newSpecPage } from '@stencil/core/testing'
import { LdCard } from '../ld-card'

describe('ld-card', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [LdCard],
      html: `<ld-card>Hello</ld-card>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders with li tag', async () => {
    const page = await newSpecPage({
      components: [LdCard],
      html: `<ld-card tag="li">Hello</ld-card>`,
    })
    expect(page.root).toMatchSnapshot()
  })

  describe('adds classes according to props', () => {
    it('adds size class', async () => {
      const page = await newSpecPage({
        components: [LdCard],
        html: `<ld-card size="sm">Hello</ld-card>`,
      })
      const card = page.root.shadowRoot.querySelector('.ld-card')
      expect(card).toHaveClass('ld-card--sm')
    })

    it('adds shadow class', async () => {
      const page = await newSpecPage({
        components: [LdCard],
        html: `<ld-card shadow="sticky">Hello</ld-card>`,
      })
      const card = page.root.shadowRoot.querySelector('.ld-card')
      expect(card).toHaveClass('ld-card--sticky')
    })

    it('adds interactive class', async () => {
      const page = await newSpecPage({
        components: [LdCard],
        html: `<ld-card interactive>Hello</ld-card>`,
      })
      const card = page.root.shadowRoot.querySelector('.ld-card')
      expect(card).toHaveClass('ld-card--interactive')
    })
  })
})
