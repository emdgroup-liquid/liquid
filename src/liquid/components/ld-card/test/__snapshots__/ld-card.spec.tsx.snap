// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-card renders 1`] = `
<ld-card class="ld-card ld-card--stacked">
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  Hello
</ld-card>
`;
